modules/core/src/smithy4s/schema/MapTag.scala

```diff
@@ -0,0 +1,71 @@
+/*
+ *  Copyright 2021-2025 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s
+package schema
+
+import scala.collection.immutable.ListMap
+
+trait MapTag[C[_, _]] {
+  def name: String
+
+  def iterator[K, V](c: C[K, V]): Iterator[(K, V)]
+  def toScalaMap[K, V](c: C[K, V]): Map[K, V] = iterator(c).toMap
+  def build[K, V](put: (((K, V)) => Unit) => Unit): C[K, V]
+  def build[K, V](preserveOrder: Boolean)(
```

Comment:
not sure if preserve order belongs here as part of the `MapLike` typeclass. I don't think it has much to do with the capabilities that a Map interface is providing. I would have a `ScalaListMapTag` or whatever we want to call it as a different implementation entirely

Reply:
So I played around with removing the `preserveOrder` argument and going with a `ScalaListMapTag`, but the problem is that since the MapTag is paramatized on `C` it's hard to change that tag to an ordered one in the code. Seems that we would have to add a `@preserveOrder` smithy trait to denote that a Map should preserve the order of keys. 

We would have to update `SchemaVisitorJCodec` to not take in the `preserveMapOrder` boolean as an argument and instead rely on hints. We will also have to update the dynamic module meta model to have the `@preserveOrder` trait for String enums and Int enums since they get serialized to a map.

Reply:
Yeah I think in the sense that the difference between a set and a list in smithy is denoted using `@uniqueItems` it would make sense that a similar "constraint" would exist at the protocol level for ordered maps, if we support it at all (which we kind of have to given the state of the dynamic module). As we talked about earlier, I think if ordering matters to a service in the request body, then the client ought to know that when it constructs said body. Similarly, if it matters to the client in a response, the service ought to take that into account. So it at least _feels_ like a protocol-level concern...

Reply:
I've created PR https://github.com/disneystreaming/smithy4s/pull/1820, which adds a `preserveKeyOrder` smithy trait and updates the codegen to pick the correct `MapTag` based if the trait is present or not.